asf_search/search/search_generator.py

```diff
@@ -78,78 +78,78 @@ def search_generator(
-
-        if maxResults != None:
-            last_page = ASFSearchResults(hits[:min(maxResults - count, len(hits))], opts=opts)
-            count += len(last_page)
-            
-            if count == maxResults:
-                last_page.searchComplete = True
-                yield last_page
-                return
-            else:
-                yield last_page
-        else:
-            count += len(hits)
-            yield ASFSearchResults(hits, opts=opts)
-
-        while('CMR-Search-After' in response.headers):
-            opts.session.headers.update({'CMR-Search-After': response.headers['CMR-Search-After']})
-
+        cmr_search_after_header = ""
+        subquery_count = 0
+        
+        while(cmr_search_after_header is not None):
             try:
-                response = get_page(session=opts.session, url=url, translated_opts=translated_opts, search_opts=query)
-            except ASFSearchError as e:
+                items, subquery_max_results, cmr_search_after_header = query_cmr(opts.session, url, translated_opts, subquery_count)
+            except (ASFSearchError, CMRIncompleteError) as e:
                 message = str(e)
                 logging.error(message)
                 report_search_error(query, message)
                 opts.session.headers.pop('CMR-Search-After', None)
                 return
-
-            hits = [ASFProduct(f, session=query.session) for f in response.json()['items']]
-
-            if len(hits):
-                if maxResults != None:
-                    last_page = ASFSearchResults(hits[:min(maxResults - count, len(hits))], opts=opts)
-                    count += len(last_page)
-                    # results.extend(hits[:min(maxResults - len(results), len(hits))])
-                    if count == maxResults:
-                        last_page.searchComplete = True
-                        yield last_page
-                        return
-                    else:
-                        yield last_page
+            
+            opts.session.headers.update({'CMR-Search-After': cmr_search_after_header})
+            total, last_page = process_page(items, maxResults, subquery_max_results, total, subquery_count, opts)
+            subquery_count += len(last_page)
+            last_page.searchComplete = subquery_count == subquery_max_results or total == maxResults
+            yield last_page
+            
+            if last_page.searchComplete:
+                if subquery_count == subquery_max_results and maxResults is None: # end of subquery
+                    cmr_search_after_header = None
+                elif total == maxResults:
+                    opts.session.headers.pop('CMR-Search-After', None)
+                    return
                 else:
-                    count += len(hits)
-                    yield ASFSearchResults(hits, opts=opts)
-
+                    cmr_search_after_header = None
```

Comment:
The first and last block here is throwing me off, they do the same thing? Would something like this work?

```python
if total == maxResutls:
    opts.session.headers.pop('CMR-Search-After', None)
    return
else:
    cmr_search_after_header = None
```

Reply:
Oh you're right!
I thought something like `maxResults == total` would throw if `maxResults` was `None` but nope!